lib/src/dart.dart

```diff
@@ -23,53 +20,26 @@ import 'io.dart';
 import 'log.dart' as log;
 
 class AnalysisContextManager {
-  /// The map from a context root directory to to the context.
-  final Map<String, AnalysisContext> _contexts = {};
-
-  /// Ensure that there are analysis contexts for the directory with the
-  /// given [path]. If any previously added root covers the [path], keep
-  /// the previously created analysis context.
-  ///
-  /// This method does not discover analysis roots "up", it only looks down
-  /// the given [path]. It is expected that the client knows analysis roots
-  /// in advance. Pub does know, it is the packages it works with.
-  void createContextsForDirectory(String path) {
-    path = p.normalize(p.absolute(path));
-
-    // We add all contexts below the given directory.
-    // So, children contexts must also have been added.
-    if (_contexts.containsKey(path)) {
-      return;
-    }
-
-    // Overwrite the analysis_options.yaml to avoid loading the file included
-    // in the package, as this may result in some files not being analyzed.
-    final resourceProvider =
-        OverlayResourceProvider(PhysicalResourceProvider.INSTANCE);
-    resourceProvider.setOverlay(
-      p.join(path, 'analysis_options.yaml'),
-      content: '',
-      modificationStamp: 0,
-    );
-
-    var contextCollection = AnalysisContextCollection(
-      includedPaths: [path],
-      resourceProvider: resourceProvider,
-      sdkPath: getSdkPath(),
-    );
-
-    // Add new contexts for the given path.
-    for (var analysisContext in contextCollection.contexts) {
-      var contextRootPath = analysisContext.contextRoot.root.path;
-
-      // If there is already a context for this context root path, keep it.
-      if (_contexts.containsKey(contextRootPath)) {
-        continue;
-      }
-
-      _contexts[contextRootPath] = analysisContext;
-    }
-  }
+  static final sessions = <String, AnalysisContextManager>{};
+
+  final String packagePath;
+  final AnalysisSession _session;
+
+  factory AnalysisContextManager(String packagePath) => sessions.putIfAbsent(
+      packagePath, () => AnalysisContextManager._(packagePath));
+
+  AnalysisContextManager._(this.packagePath)
+      : _session = ContextBuilder()
+            .createContext(
+              contextRoot: ContextLocator().locateRoots(
+                includedPaths: [p.absolute(p.normalize(packagePath))],
+                optionsFile:
+                    // We don't want to take 'analysis_options.yaml' files into
+                    // account. So we replace it with an empty file.
+                    Platform.isWindows ? r'C:\NUL' : '/dev/null',
```

Comment:
that works everywhere?

Reply:
At least on our CI bots...

Reply:
Reading from a file called NUL on windows gives an empty file: https://ss64.com/nt/nul.html I use `C:\NUL` because the analyzer insists on absolute paths (WHY?)

Reading from /dev/null on posix gives an empty file: https://linuxhint.com/what_is_dev_null/ (near the bottom).